pkg/sync/rpc/server.go

```diff
@@ -329,8 +329,9 @@ func (*SyncAgentServer) FileRename(ctx context.Context, req *ptypes.FileRenameRe
 
 func (s *SyncAgentServer) FileSend(ctx context.Context, req *ptypes.FileSendRequest) (*empty.Empty, error) {
 	address := net.JoinHostPort(req.Host, strconv.Itoa(int(req.Port)))
+	directIO := false
```

Comment:
By default, it should be true

Reply:
When it set as true, it will cause 12 integration tests to fail.  All have failure message similar as below:

```
time="2021-02-24T08:33:54Z" level=fatal msg="Error running add replica command: failed to prepare rebuilding replica tcp://localhost:10015 for volume localhost:30001: rpc error: code = Unknown desc = failed to send file volume-head-002.img.meta to localhost:10018: rpc error: code = Unknown desc = syncDataInterval [       0:       1](  1) failed, err: failed to get checksums for client and server, server checksum error: <nil>, client checksum error: File to read interval:[       0:       1](  1), error: read volume-head-002.img.meta: invalid argument"
```

Reply:
That's unexpected... We may need to modify the related tests and make sure the replica size is the multiple of 512Byte

Reply:
May I know the context of using 512byte? Seems like some place using 512byte and other using 4k.

```
Blocks int64 = 4 << 10 // 4k => github.com/longhorn/longhorn-engine/vendor/github.com/longhorn/sparse-tools/sparse/layout.go
const defaultSectorSize = 512 // => github.com/longhorn/longhorn-engine/pkg/backing/backing.go
defaultSectorSize = 4096 // => github.com/longhorn/longhorn-engine/pkg/replica/replica.go
```

Reply:
512 is the sector size of qcow2 files, which we might use for backing images. In general, we want to be 4k aligned for the size. We need the alignment for the directIO read/write.

In the error message above, it's about `.img.meta` which is a metadata file, so it's not 512 nor 4k aligned. So for those files, directIO should be disabled. Not sure we are having issues there.

Reply:
It seems that DirectIO requires `fileSize > 4096` intead of `filesize % 4096`, so I have made modifications as following:

```
if `directIO == true` and `fileSize >= 4096`, then server/client use directIO.
if `directIO == true` and `fileSize < 4096`, then server/client do not use directIO.
if `directIO == false` and `fileSize >= 4096`, then server/client do not use directIO.
if `directIO == false` and `fileSize < 0`, then server/client do not use directIO.
```

And add unit test as in https://github.com/longhorn/sparse-tools/pull/70

Reply:
Is `fileSize >= 4096` enough for DirectIO? Is `fileSize % 512 == 0` unnecessary?

Reply:
I have created a few [test cases](https://github.com/longhorn/sparse-tools/pull/70/files#diff-5f8226886adcb74f851e23e6bcdf39310d3e63cdc1feb8f6d85895d00ca52502).  After running the test with different configurations, It seems that error only occurred when file size is less than 4096 KB.  When I test what test case (4096 KB + 256 KB) for file size, `directIO` is able to function properly and `checkSparseFiles` is also passed.

Reply:
4MB + 256KB? Or you mean 4MB + 256B? If it's 256KB, this mathes the 512B alignment...

Reply:
OK. It seems that the sector size depends on the filesystem type. For EXT4, the default sector size is 512B rather than 4KB
```
root@shuo-cluster-0-worker-1:~# fdisk -l
...

Disk /dev/sda: 79.5 GiB, 85362475008 bytes, 166723584 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes

...
```

I am not sure why we use 4K alignment in longhorn sparse file repo. @yasker Do you have any contexts?

Right now can you try 2 tests with directIO forcibly enabled first:
1. R/W a file that is 512B alignment but not 4KB alignment.
2. R/W a file that is neither 512B alignment nor 4KB alignment.

Reply:
I think the reason we use 4k is to improve write speed.

Reply:
I created files 
```
rm -rf /tmp/test*.tmp
dd iflag=fullblock if=/dev/random of=/tmp/test4096KB.tmp bs=4k count=25600 && sync
dd iflag=fullblock if=/dev/random of=/tmp/test512KB.tmp bs=512 count=204800 && sync
dd iflag=fullblock if=/dev/random of=/tmp/test256KB.tmp bs=256 count=409600 && sync
ls -alh /tmp/test*.tmp
```
start daemon
```
FILE=/tmp/file256KB.tmp # /tmp/file.qcow2 | /tmp/file4096KB.tmp | /tmp/file512KB.tmp | /tmp/file256KB.tmp
cd ~/src/github.com/cclhsu/sparse-tools/bin
rm -rf ${FILE}
./ssync -verbose -daemon ${FILE}
```
start client
```
~/src/github.com/cclhsu/sparse-tools/bin
FILE=/tmp/test256KB.tmp # ~/Documents/myImages/kvm/ubuntu/ubuntu-20.04-server-cloudimg-arm64.img | /tmp/test4096KB.tmp | /tmp/test512KB.tmp | /tmp/test256KB.tmp
./ssync -host 0.0.0.0 ${FILE}
```
It seems that diff and size are the same. For Qcow2 image, it is checked with command `qemu-img check` with not errors.

Reply:
> I am not sure why we use 4K alignment in longhorn sparse file repo. @yasker Do you have any contexts?

@yasker do you know any information on this?

Reply:
Yes, 4k was used before introducing the backing image feature, because then we can save 8x times of space for tracking a block/sector(each sector needs to be tracked inside memory for us to know where to read/write), also we might increase the performance in that way. However with qcow2 it was 512B aligned, so IIRC we basically switch to 512 later.

I am fine with either now. Stick with 4k for the size alignment is probably easier. But internally, at least with qcow2 backing image, we have to use 512B as sector size inside the replica/engine.

@shuo-wu For directIO, every read/write have to be aligned with the block size underneath. So it's not `fileSize > 4k`, but `filesize % 4k == 0` or `filesize % 512 == 0`.

@cclhsu I am not sure why we have files that are less than 4k but required for directIO. I would rather stick to 4k alignment for all directIO needed files if possible. The replica size should be always multiples of 4k IIRC.

Reply:
The reason for the failure of

```
time="2021-02-24T08:33:54Z" level=fatal msg="Error running add replica command: failed to prepare rebuilding replica tcp://localhost:10015 for volume localhost:30001: rpc error: code = Unknown desc = failed to send file volume-head-002.img.meta to localhost:10018: rpc error: code = Unknown desc = syncDataInterval [       0:       1](  1) failed, err: failed to get checksums for client and server, server checksum error: <nil>, client checksum error: File to read interval:[       0:       1](  1), error: read volume-head-002.img.meta: invalid argument"
```

is this is a metadata file. Metadata file (ending with `.meta`) isn't designed to be transferred using directIO. The directIO is used to persist data to the disk for the images (files ending with `.img`), not metadata files.